chromium_src/base/debug/alias.h

```diff
@@ -0,0 +1,43 @@
+/* Copyright (c) 2024 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#ifndef BRAVE_CHROMIUM_SRC_BASE_DEBUG_ALIAS_H_
+#define BRAVE_CHROMIUM_SRC_BASE_DEBUG_ALIAS_H_
+
+#include <stdint.h>
+#include <string.h>
+
+#include "base/memory/raw_ptr_exclusion.h"
+
+#include "src/base/debug/alias.h"  // IWYU pragma: export
+
+namespace base {
+namespace debug {
+
+template <typename T>
+class StackObjectCopy {
+ public:
+  explicit StackObjectCopy(const T* original) {
+    if (original) {
+      memcpy(static_cast<void*>(buffer_), static_cast<const void*>(original),
+             sizeof(T));
+    } else {
+      memset(buffer_, 0, sizeof(T));
+    }
+  }
+
+  StackObjectCopy(const StackObjectCopy&) = delete;
+  StackObjectCopy& operator=(const StackObjectCopy&) = delete;
+
+ private:
+  alignas(T) uint8_t buffer_[sizeof(T)];
+  RAW_PTR_EXCLUSION const T* const typed_view_ =
+      reinterpret_cast<const T*>(buffer_);
```

Comment:
There are two issues I'm having with this class as it is. First we are doing something like this:

```cxx
  base::debug::StackObjectCopy<ExecutionContext> context_copy(context);
  base::debug::Alias(&context_copy);
  NOTREACHED() << "Unhandled ExecutionContext type"
```
However `blink::ExecutionContext` is partially an abstract class, so we are slicing off the extended data from the type. I'm not sure how that interacts with error reporting, but it certainly will omit relevant data.

The other problem is this:

```
  RAW_PTR_EXCLUSION const T* const typed_view_ =
      reinterpret_cast<const T*>(buffer_);
```

So we `reinterpret_cast` `buffer_` as a `const T*`, however the data pointed by `buffer_` is not `T`, but rather a copy of the bytes. As far as I understand reading this data as T constitutes Undefined Behaviour.

Reply:
> `blink::ExecutionContext` is partially an abstract class, so we are slicing off the extended data from the type. I'm not sure how that interacts with error reporting, but it certainly will omit relevant data.

that is correct. yeah. This helper is added to see the actual class type and its base members in crash dumps. This is not intended to determine the full class size and keep everything from it.

> we reinterpret_cast buffer_ as a const T*, however the data pointed by buffer_ is not T, but rather a copy of the bytes. As far as I understand reading this data as T constitutes Undefined Behaviour.

 The `typed_view_` variable is only for the purpose to see it in the debugger.

Reply:
I guess we need to determine though if that would be undefined behaviour. If it is, we are in a very thin ice with this, and it could be very unreliable. I'm looking around to see what I can find on this.